File: frontend/recetario/lib/data/models/tema.dart
import 'material.dart';
import 'tarea.dart';

class Tema {
  final String id;
  final String cursoId;
  final String titulo;
  final String? descripcion;
  final int orden;
  final DateTime? fechaDesbloqueo;
  final bool activo;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<Material>? materiales;
  final List<Tarea>? tareas;

  Tema({
    required this.id,
    required this.cursoId,
    required this.titulo,
    this.descripcion,
    required this.orden,
    this.fechaDesbloqueo,
    required this.activo,
    required this.createdAt,
    required this.updatedAt,
    this.materiales,
    this.tareas,
  });

  factory Tema.fromJson(Map<String, dynamic> json) {
    return Tema(
      id: json['id'],
      cursoId: json['curso_id'],
      titulo: json['titulo'],
      descripcion: json['descripcion'],
      orden: json['orden'],
      fechaDesbloqueo: json['fecha_desbloqueo'] != null
          ? DateTime.parse(json['fecha_desbloqueo'])
          : null,
      activo: json['activo'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      materiales: json['materiales'] != null
          ? (json['materiales'] as List)
              .map((m) => Material.fromJson(m))
              .toList()
          : null,
      tareas: json['tareas'] != null
          ? (json['tareas'] as List).map((t) => Tarea.fromJson(t)).toList()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'curso_id': cursoId,
      'titulo': titulo,
      'descripcion': descripcion,
      'orden': orden,
      'fecha_desbloqueo': fechaDesbloqueo?.toIso8601String(),
    };
  }

  bool get estaBloqueado {
    if (fechaDesbloqueo == null) return false;
    return DateTime.now().isBefore(fechaDesbloqueo!);
  }
}

File: backend/internal/models/tema.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Tema struct {
	ID              uuid.UUID  `json:"id" db:"id"`
	CursoID         uuid.UUID  `json:"curso_id" db:"curso_id"`
	Titulo          string     `json:"titulo" db:"titulo"`
	Descripcion     *string    `json:"descripcion,omitempty" db:"descripcion"`
	Orden           int        `json:"orden" db:"orden"`
	FechaDesbloqueo *time.Time `json:"fecha_desbloqueo,omitempty" db:"fecha_desbloqueo"`
	Activo          bool       `json:"activo" db:"activo"`
	CreatedAt       time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at" db:"updated_at"`

	// Relaciones
	Materiales []Material `json:"materiales,omitempty"`
	Tareas     []Tarea    `json:"tareas,omitempty"`
}

type CreateTemaRequest struct {
	CursoID         uuid.UUID  `json:"curso_id" binding:"required"`
	Titulo          string     `json:"titulo" binding:"required"`
	Descripcion     *string    `json:"descripcion"`
	Orden           int        `json:"orden" binding:"required,min=1"`
	FechaDesbloqueo *time.Time `json:"fecha_desbloqueo"`
}

type UpdateTemaRequest struct {
	Titulo          *string    `json:"titulo"`
	Descripcion     *string    `json:"descripcion"`
	Orden           *int       `json:"orden" binding:"omitempty,min=1"`
	FechaDesbloqueo *time.Time `json:"fecha_desbloqueo"`
	Activo          *bool      `json:"activo"`
}
